core/src/main/java/com/oath/oak/EntrySet.java

```diff
@@ -0,0 +1,951 @@
+         * {@code version > INVALID_VERSION} if the insertion was completed.
+         */
+        int version;
+        /* key reference, build differently than value reference
+          * TODO: eliminate parsing key and value references by any entity other than EntrySet */
+        final long keyReference;
+
+        LookUp(Slice valueSlice, long valueReference, int entryIndex, int version,
+            long keyReference, boolean isFinalizeDeleteNeeded) {
+            this.valueSlice = valueSlice;
+            this.valueReference = valueReference;
+            this.entryIndex = entryIndex;
+            this.version = version;
+            this.keyReference = keyReference;
+            this.isFinalizeDeletionNeeded = isFinalizeDeleteNeeded;
+        }
+    }
+
+
+    /********************************************************************************************/
+    /*---------------- Methods for setting and getting specific entry's field ------------------*/
+    // we use term intIdx for the integer's index inside the entries array (referred as a set of integers),
+    // we use term entryIdx for the index of entry array (referred as a set of entries)
+
+    private int entryIdx2intIdx(int entryIdx) {
+        if (entryIdx == 0) { // assumed it is a head pointer access
+            return 0;
+        }
+        return ((entryIdx-1) * FIELDS) + FIRST_ITEM;
+    }
+
+    /**
+     * getEntryArrayFieldInt gets the integer field of entry at specified offset for given
+     * start of the entry index in the entry array.
+     * The field is read atomically as it is a machine word, however the concurrency of the
+     * mostly updated value is not ensured as no memory fence is issued.
+     */
+    private int getEntryArrayFieldInt(int intFieldIdx, OFFSET offset) {
+        switch (offset) {
+        case KEY_LENGTH:
+            // return two low bytes of the key length index int
+            return (entries[intFieldIdx + offset.value] & KEY_LENGTH_MASK);
+        case KEY_BLOCK:
+            // offset must be OFFSET_KEY_BLOCK, return 2 high bytes of the int inside key length
+            // right-shift force, fill empty with zeroes
+            return (entries[intFieldIdx + offset.value] >>> KEY_BLOCK_SHIFT);
+        case VALUE_LENGTH:
+            return (entries[intFieldIdx + offset.value] & VALUE_LENGTH_MASK);
+        case VALUE_BLOCK:
+            return (entries[intFieldIdx + offset.value] >>> VALUE_BLOCK_SHIFT);
+        default:
+            return entries[intFieldIdx + offset.value];
+        }
+    }
+
+    /**
+     * getEntryArrayFieldLong atomically reads two integers field of the entries array.
+     * Should be used with OFFSET.VALUE_REFERENCE and OFFSET.KEY_REFERENCE
+     * The concurrency is ensured due to memory fence as part of "volatile"
+     */
+    private long getEntryArrayFieldLong(int intStartFieldIdx, OFFSET offset) {
+        long arrayOffset =
```

Comment:
if not all OFFSET values are allowed - should you check ant throw exception?

Reply:
fixed